In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
from tqdm import tqdm_notebook

## Train과 Test 로드 후 병합
- Train Id : 1 ~ 1751
- Test Id : 1752 ~ 2919

In [2]:
    # train,test 데이터 로드
    train = pd.read_csv('train.csv')    
    test = pd.read_csv('test.csv')

    # SalePrice를 별도로 분리
    # inplace 속성으로 drop결과를 바로 적용
    targets = train.SalePrice
    train.drop(['SalePrice'], axis = 1, inplace=True)
    
    # 두 개의 데이터 병합
    # ignore_index 속성으로 train,test의 인덱스를 순차적으로 만듬
    combined = train.append(test,ignore_index=True)

    combined

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          20       RL          NaN    21695   Pave   NaN      IR1   
1        2          20       RL          NaN    10530   Pave   NaN      IR1   
2        3          20       RL         88.0    12803   Pave   NaN      IR1   
3        4          60       FV        114.0     8314   Pave  Pave      IR1   
4        5          20       RL         78.0    10335   Pave   NaN      IR1   
5        6          90       RL         55.0    12640   Pave   NaN      IR1   
6        7          20       RL         85.0    11049   Pave   NaN      Reg   
7        8          20       RL         60.0     6600   Pave   NaN      Reg   
8        9          80       RL         65.0     6305   Pave   NaN      Reg   
9       10          80       RL         55.0     7700   Pave   NaN      Reg   
10      11          20       RL          NaN     8475   Pave   NaN      IR1   
11      12          20       RL        100.0    10175   Pave   NaN      IR1   
12      13          60       FV          NaN     7500   Pave   NaN      Reg   
13      14          20       RL         60.0    10434   Pave   NaN      Reg   
14      15          70       RM         60.0     3600   Pave  Grvl      Reg   
15      16         190       RL         70.0    13587   Pave   NaN      Reg   
16      17          20       RL          NaN    26142   Pave   NaN      IR1   
17      18          30       RM         50.0     5000   Pave   NaN      Reg   
18      19          70       RM         51.0     6120   Pave   NaN      Reg   
19      20          60       RL         92.0    15922   Pave   NaN      IR1   
20      21          60       RL         64.0     8333   Pave   NaN      Reg   
21      22          60       RL         68.0     8846   Pave   NaN      Reg   
22      23          20       RL         85.0    11900   Pave   NaN      Reg   
23      24          70       RL         75.0    13125   Pave   NaN      Reg   
24      25          60       RL         62.0     7162   Pave   NaN      Reg   
25      26          20       RL          NaN    13680   Pave   NaN      IR1   
26      27          80       RL          NaN     7750   Pave   NaN      IR1   
27      28          20       RL         76.0    11355   Pave   NaN      IR1   
28      29         190       RM         60.0    10800   Pave  Pave      Reg   
29      30         160       RM         24.0     2522   Pave   NaN      Reg   
...    ...         ...      ...          ...      ...    ...   ...      ...   
2889  2890          20       RL         80.0     8816   Pave   NaN      Reg   
2890  2891          70       RM         57.0     9639   Pave   NaN      Reg   
2891  2892          60       FV         84.0    10207   Pave   NaN      Reg   
2892  2893          20       RL          NaN    12493   Pave   NaN      IR1   
2893  2894         120       RL         61.0     7658   Pave   NaN      Reg   
2894  2895          50       RL         67.0    12354   Pave  Grvl      Reg   
2895  2896          90       RL         91.0    11643   Pave   NaN      Reg   
2896  2897         120       RL          NaN     3696   Pave   NaN      Reg   
2897  2898          20       RL         64.0     7697   Pave   NaN      Reg   
2898  2899          20       RL         80.0    11900   Pave   NaN      IR1   
2899  2900          20       RL         91.0    11024   Pave   NaN      IR1   
2900  2901          90       RL          NaN     7424   Pave   NaN      IR1   
2901  2902          50       RL         60.0     7200   Pave   NaN      Reg   
2902  2903          60       RL          NaN     8068   Pave   NaN      IR1   
2903  2904          60       RL         90.0     9900   Pave   NaN      Reg   
2904  2905          20       RL        104.0    11361   Pave   NaN      Reg   
2905  2906          60       RL         84.0    14260   Pave   NaN      IR1   
2906  2907          30       RL          NaN    12342   Pave   NaN      IR1   
2907  2908          60       RL          NaN    11613   Pa

In [3]:
combined.shape

(2919, 80)

In [4]:
combined.isnull().sum()

Id                  0
MSSubClass          0
MSZoning            4
LotFrontage       486
LotArea             0
Street              0
Alley            2721
LotShape            0
LandContour         0
Utilities           2
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
OverallQual         0
OverallCond         0
YearBuilt           0
YearRemodAdd        0
RoofStyle           0
RoofMatl            0
Exterior1st         1
Exterior2nd         1
MasVnrType         24
MasVnrArea         23
ExterQual           0
ExterCond           0
Foundation          0
                 ... 
HalfBath            0
BedroomAbvGr        0
KitchenAbvGr        0
KitchenQual         1
TotRmsAbvGrd        0
Functional          2
Fireplaces          0
FireplaceQu      1420
GarageType        157
GarageYrBlt       159
GarageFinish      159
GarageCars          1
GarageArea          1
GarageQual        159
GarageCond

### 컬럼 수가 80개라 다 안보여서 a, b, c로 자름
- 결측치 찾기

In [5]:
a_combined = combined.iloc[:, :30]
b_combined = combined.iloc[:, 30:60]
c_combined = combined.iloc[:, 60:]

In [6]:
a_combined.isnull().sum()

Id                 0
MSSubClass         0
MSZoning           4
LotFrontage      486
LotArea            0
Street             0
Alley           2721
LotShape           0
LandContour        0
Utilities          2
LotConfig          0
LandSlope          0
Neighborhood       0
Condition1         0
Condition2         0
BldgType           0
HouseStyle         0
OverallQual        0
OverallCond        0
YearBuilt          0
YearRemodAdd       0
RoofStyle          0
RoofMatl           0
Exterior1st        1
Exterior2nd        1
MasVnrType        24
MasVnrArea        23
ExterQual          0
ExterCond          0
Foundation         0
dtype: int64

In [7]:
b_combined.isnull().sum()

BsmtQual          81
BsmtCond          82
BsmtExposure      82
BsmtFinType1      79
BsmtFinSF1         1
BsmtFinType2      80
BsmtFinSF2         1
BsmtUnfSF          1
TotalBsmtSF        1
Heating            0
HeatingQC          0
CentralAir         0
Electrical         1
1stFlrSF           0
2ndFlrSF           0
LowQualFinSF       0
GrLivArea          0
BsmtFullBath       2
BsmtHalfBath       2
FullBath           0
HalfBath           0
BedroomAbvGr       0
KitchenAbvGr       0
KitchenQual        1
TotRmsAbvGrd       0
Functional         2
Fireplaces         0
FireplaceQu     1420
GarageType       157
GarageYrBlt      159
dtype: int64

In [8]:
c_combined.isnull().sum()

GarageFinish      159
GarageCars          1
GarageArea          1
GarageQual        159
GarageCond        159
PavedDrive          0
WoodDeckSF          0
OpenPorchSF         0
EnclosedPorch       0
3SsnPorch           0
ScreenPorch         0
PoolArea            0
PoolQC           2909
Fence            2348
MiscFeature      2814
MiscVal             0
MoSold              0
YrSold              0
SaleType            1
SaleCondition       0
dtype: int64

In [9]:
a_combined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2919 entries, 0 to 2918
Data columns (total 30 columns):
Id              2919 non-null int64
MSSubClass      2919 non-null int64
MSZoning        2915 non-null object
LotFrontage     2433 non-null float64
LotArea         2919 non-null int64
Street          2919 non-null object
Alley           198 non-null object
LotShape        2919 non-null object
LandContour     2919 non-null object
Utilities       2917 non-null object
LotConfig       2919 non-null object
LandSlope       2919 non-null object
Neighborhood    2919 non-null object
Condition1      2919 non-null object
Condition2      2919 non-null object
BldgType        2919 non-null object
HouseStyle      2919 non-null object
OverallQual     2919 non-null int64
OverallCond     2919 non-null int64
YearBuilt       2919 non-null int64
YearRemodAdd    2919 non-null int64
RoofStyle       2919 non-null object
RoofMatl        2919 non-null object
Exterior1st     2918 non-null object
Exterior2nd 

In [10]:
b_combined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2919 entries, 0 to 2918
Data columns (total 30 columns):
BsmtQual        2838 non-null object
BsmtCond        2837 non-null object
BsmtExposure    2837 non-null object
BsmtFinType1    2840 non-null object
BsmtFinSF1      2918 non-null float64
BsmtFinType2    2839 non-null object
BsmtFinSF2      2918 non-null float64
BsmtUnfSF       2918 non-null float64
TotalBsmtSF     2918 non-null float64
Heating         2919 non-null object
HeatingQC       2919 non-null object
CentralAir      2919 non-null object
Electrical      2918 non-null object
1stFlrSF        2919 non-null int64
2ndFlrSF        2919 non-null int64
LowQualFinSF    2919 non-null int64
GrLivArea       2919 non-null int64
BsmtFullBath    2917 non-null float64
BsmtHalfBath    2917 non-null float64
FullBath        2919 non-null int64
HalfBath        2919 non-null int64
BedroomAbvGr    2919 non-null int64
KitchenAbvGr    2919 non-null int64
KitchenQual     2918 non-null object
TotRmsA

In [11]:
c_combined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2919 entries, 0 to 2918
Data columns (total 20 columns):
GarageFinish     2760 non-null object
GarageCars       2918 non-null float64
GarageArea       2918 non-null float64
GarageQual       2760 non-null object
GarageCond       2760 non-null object
PavedDrive       2919 non-null object
WoodDeckSF       2919 non-null int64
OpenPorchSF      2919 non-null int64
EnclosedPorch    2919 non-null int64
3SsnPorch        2919 non-null int64
ScreenPorch      2919 non-null int64
PoolArea         2919 non-null int64
PoolQC           10 non-null object
Fence            571 non-null object
MiscFeature      105 non-null object
MiscVal          2919 non-null int64
MoSold           2919 non-null int64
YrSold           2919 non-null int64
SaleType         2918 non-null object
SaleCondition    2919 non-null object
dtypes: float64(2), int64(9), object(9)
memory usage: 456.2+ KB


### 결측치있는 컬럼들

MSZoning           4
LotFrontage      486
Alley           2721
Utilities          2
Exterior1st        1
Exterior2nd        1
MasVnrType        24
MasVnrArea        23
BsmtQual          81
BsmtCond          82
BsmtExposure      82
BsmtFinType1      79
BsmtFinSF1         1
BsmtFinType2      80
BsmtFinSF2         1
BsmtUnfSF          1
TotalBsmtSF        1
Electrical         1
BsmtFullBath       2
BsmtHalfBath       2
KitchenQual        1
Functional         2
FireplaceQu     1420
GarageType       157
GarageYrBlt      159
GarageFinish      159
GarageCars          1
GarageArea          1
GarageQual        159
GarageCond        159
PoolQC           2909
Fence            2348
MiscFeature      2814
SaleType            1

### 결측치가 1000개 이상 컬럼들

"Alley", "FireplaceQu", "PoolQC", "Fence", "MiscFeature"

In [12]:
# "Alley", "FireplaceQu", "PoolQC", "Fence", "MiscFeature" 특성 삭제
combined.drop('Alley', axis=1, inplace=True)
combined.drop('FireplaceQu', axis=1, inplace=True)
combined.drop('PoolQC', axis=1, inplace=True)
combined.drop('Fence', axis=1, inplace=True)
combined.drop('MiscFeature', axis=1, inplace=True)

### 나머지 결측치 있는 컬럼들 : 29

"MSZoning"  "LotFrontage" "Utilities"  "Exterior1st"  "Exterior2nd"  "MasVnrType"  "MasVnrArea"  "BsmtQual"  "BsmtCond"  "BsmtExposure"  "BsmtFinType1"  "BsmtFinSF1"  "BsmtFinType2"  "BsmtFinSF2"  "BsmtUnfSF"  "TotalBsmtSF"  "Electrical"  "BsmtFullBath"  "BsmtHalfBath"  "KitchenQual"  "Functional" "GarageType" "GarageYrBlt"  "GarageFinish"  "GarageCars"  "GarageArea"  "GarageQual"  "GarageCond"  "SaleType" 

### 문자, 숫자 데이터 분류

"MSZoning", "Utilities", "Exterior1st", "Exterior2nd" , "MasVnrType", "BsmtQual", "BsmtCond", "BsmtExposure", "BsmtFinType1", "BsmtFinType2", "Electrical",  "KitchenQual" , "Functional", "GarageType", "GarageFinish", "GarageQual", "GarageCond", "SaleType"
- 문자 : 18

"LotFrontage", "MasVnrArea", "BsmtFinSF1", "BsmtFinSF2", "BsmtUnfSF", "TotalBsmtSF", "BsmtFullBath", "BsmtHalfBath", "GarageYrBlt", "GarageCars", "GarageArea"
- 숫자 : 11

## 문자데이터 결측치 채우기

In [13]:
combined.iloc[:891].MSZoning.value_counts()          # 4
# 결측치 처리
combined.MSZoning.fillna('RL', inplace=True)

In [14]:
combined.iloc[:891].Utilities.value_counts()        # 2
combined.Utilities.fillna('AllPub', inplace=True)

In [15]:
combined.iloc[:891].Exterior1st.value_counts()      # 1
combined.Exterior1st.fillna('VinylSd', inplace=True)

In [16]:
combined.iloc[:891].Exterior2nd.value_counts()      # 1
combined.Exterior2nd.fillna('VinylSd', inplace=True)

In [17]:
combined.iloc[:891].MasVnrType.value_counts()     # 24 
combined.MasVnrType.fillna('None', inplace=True)

In [18]:
combined.iloc[:891].BsmtQual.value_counts()      # 81
combined.BsmtQual.fillna('missing', inplace=True)                # missing 으로 채우기

In [19]:
combined.iloc[:891].BsmtCond.value_counts()     # 82
combined.BsmtCond.fillna('TA', inplace=True)

In [20]:
combined.iloc[:891].BsmtExposure.value_counts()    # 82
combined.BsmtExposure.fillna('No', inplace=True)

In [21]:
combined.iloc[:891].BsmtFinType1.value_counts()    # 79
combined.BsmtFinType1.fillna('missing', inplace=True)            # missing 으로 채우기

In [22]:
combined.iloc[:891].BsmtFinType2.value_counts()    # 80
combined.BsmtFinType2.fillna('Unf', inplace=True)

In [23]:
combined.iloc[:891].Electrical.value_counts()      # 1
combined.Electrical.fillna('SBrkr', inplace=True)

In [24]:
combined.iloc[:891].KitchenQual.value_counts()    # 1
combined.KitchenQual.fillna('TA', inplace=True)

In [25]:
combined.iloc[:891].Functional.value_counts()     # 2
combined.Functional.fillna('Typ', inplace=True)

In [26]:
combined.iloc[:891].GarageType.value_counts()    # 157
combined.GarageType.fillna('missing', inplace=True)            # missing 으로 채우기

In [27]:
combined.iloc[:891].GarageFinish.value_counts()  # 159
combined.GarageFinish.fillna('missing', inplace=True)         # missing 으로 채우기

In [28]:
combined.iloc[:891].GarageQual.value_counts()   # 159
combined.GarageQual.fillna('TA', inplace=True)

In [29]:
combined.iloc[:891].GarageCond.value_counts()  # 159
combined.GarageCond.fillna('TA', inplace=True)

In [30]:
combined.iloc[:891].SaleType.value_counts()  # 1
combined.SaleType.fillna('WD', inplace=True)

## 숫자 데이터 결측치 채우기

In [31]:
combined.iloc[:891].LotFrontage.value_counts()  # 486
LotFrontage_med = combined.groupby("MSSubClass")["LotFrontage"].transform('median') 
LotFrontage_med
combined["LotFrontage"]=combined["LotFrontage"].fillna(LotFrontage_med)                # 중앙값으로 채우기

In [32]:
combined.iloc[:891].MasVnrArea.value_counts()  # 23
MasVnrArea_med = combined.groupby("MSSubClass")["MasVnrArea"].transform('median') 
MasVnrArea_med
combined["MasVnrArea"]=combined["MasVnrArea"].fillna(MasVnrArea_med)                   # 중앙값으로 채우기

In [33]:
combined.iloc[:891].BsmtFinSF1.value_counts()  # 1
combined.BsmtFinSF1.fillna(0.0, inplace=True)               # 최빈값으로 채우기

In [34]:
combined.iloc[:891].BsmtFinSF2.value_counts()  # 1
combined.BsmtFinSF2.fillna(0.0, inplace=True)                # 최빈값으로 채우기

In [35]:
combined.iloc[:891].BsmtUnfSF.value_counts()  # 1
combined.BsmtUnfSF.fillna(0.0, inplace=True)                # 최빈값으로 채우기

In [36]:
combined.iloc[:891].TotalBsmtSF.value_counts()  # 1
TotalBsmtSF_med = combined.groupby("MSSubClass")["TotalBsmtSF"].transform('median') 
TotalBsmtSF_med
combined["TotalBsmtSF"]=combined["TotalBsmtSF"].fillna(TotalBsmtSF_med)                   # 중앙값으로 채우기

In [37]:
combined.iloc[:891].BsmtFullBath.value_counts()  # 2
combined.BsmtFullBath.fillna(0.0, inplace=True)                # 최빈값으로 채우기

In [38]:
combined.iloc[:891].BsmtHalfBath.value_counts()  # 2
combined.BsmtHalfBath.fillna(0.0, inplace=True)                  # 최빈값으로 채우기

In [39]:
combined.iloc[:891].GarageYrBlt.value_counts()  # 159
GarageYrBlt_med = combined.groupby("MSSubClass")["GarageYrBlt"].transform('median') 
GarageYrBlt_med
combined["GarageYrBlt"]=combined["GarageYrBlt"].fillna(GarageYrBlt_med)                   # 중앙값으로 채우기

In [40]:
combined.iloc[:891].GarageCars.value_counts()  # 1
combined.GarageCars.fillna(2.0, inplace=True)                  # 최빈값으로 채우기

In [41]:
combined.iloc[:891].GarageArea.value_counts()  # 1
combined.GarageArea.fillna(0.0, inplace=True)                  # 최빈값으로 채우기

## 결측치 문자, 숫자 다 채운 후 확인하기

In [42]:
a_combined = combined.iloc[:, :30]
b_combined = combined.iloc[:, 30:60]
c_combined = combined.iloc[:, 60:]

In [43]:
a_combined.isnull().sum()

Id              0
MSSubClass      0
MSZoning        0
LotFrontage     1
LotArea         0
Street          0
LotShape        0
LandContour     0
Utilities       0
LotConfig       0
LandSlope       0
Neighborhood    0
Condition1      0
Condition2      0
BldgType        0
HouseStyle      0
OverallQual     0
OverallCond     0
YearBuilt       0
YearRemodAdd    0
RoofStyle       0
RoofMatl        0
Exterior1st     0
Exterior2nd     0
MasVnrType      0
MasVnrArea      0
ExterQual       0
ExterCond       0
Foundation      0
BsmtQual        0
dtype: int64

In [44]:
b_combined.isnull().sum()

BsmtCond        0
BsmtExposure    0
BsmtFinType1    0
BsmtFinSF1      0
BsmtFinType2    0
BsmtFinSF2      0
BsmtUnfSF       0
TotalBsmtSF     0
Heating         0
HeatingQC       0
CentralAir      0
Electrical      0
1stFlrSF        0
2ndFlrSF        0
LowQualFinSF    0
GrLivArea       0
BsmtFullBath    0
BsmtHalfBath    0
FullBath        0
HalfBath        0
BedroomAbvGr    0
KitchenAbvGr    0
KitchenQual     0
TotRmsAbvGrd    0
Functional      0
Fireplaces      0
GarageType      0
GarageYrBlt     0
GarageFinish    0
GarageCars      0
dtype: int64

In [45]:
c_combined.isnull().sum()

GarageArea       0
GarageQual       0
GarageCond       0
PavedDrive       0
WoodDeckSF       0
OpenPorchSF      0
EnclosedPorch    0
3SsnPorch        0
ScreenPorch      0
PoolArea         0
MiscVal          0
MoSold           0
YrSold           0
SaleType         0
SaleCondition    0
dtype: int64

### 결측치 처리 후에도 남아있는 결측치 행 검색
- Train Id : 1 ~ 1751
- Test Id : 1752 ~ 2919

In [46]:
combined[combined["LotFrontage"].isnull()==True]  

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape  \
1705  1706         150       RL          NaN     1700   Pave      Reg   

     LandContour Utilities LotConfig  ... OpenPorchSF EnclosedPorch 3SsnPorch  \
1705         HLS    AllPub    Inside  ...          36             0         0   

     ScreenPorch PoolArea MiscVal  MoSold  YrSold  SaleType  SaleCondition  
1705         200        0       0       4    2006        WD         Normal  

[1 rows x 75 columns]

In [47]:
combined.drop([1705], axis=0, inplace=True)

In [48]:
a_combined = combined.iloc[:, :30]
b_combined = combined.iloc[:, 30:60]
c_combined = combined.iloc[:, 60:]

In [49]:
a_combined.isnull().sum()

Id              0
MSSubClass      0
MSZoning        0
LotFrontage     0
LotArea         0
Street          0
LotShape        0
LandContour     0
Utilities       0
LotConfig       0
LandSlope       0
Neighborhood    0
Condition1      0
Condition2      0
BldgType        0
HouseStyle      0
OverallQual     0
OverallCond     0
YearBuilt       0
YearRemodAdd    0
RoofStyle       0
RoofMatl        0
Exterior1st     0
Exterior2nd     0
MasVnrType      0
MasVnrArea      0
ExterQual       0
ExterCond       0
Foundation      0
BsmtQual        0
dtype: int64

## combined 를 다시 Train과 Test 로 나눔
- combined에서 train 데이터 1개 행을 드롭.
- 자동생성 인덱스로는 0 ~ 1750 / 1751 ~
- Train Id : 1 ~ 1751
- Test Id : 1752 ~ 2919

In [50]:
train[1749:]

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
1749  1750          20       RL          NaN    17597   Pave   NaN      Reg   
1750  1751          80       RL         74.0     9620   Pave   NaN      Reg   

     LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence  \
1749         Bnk    AllPub  ...           0        0    NaN  GdPrv   
1750         Lvl    AllPub  ...           0        0    NaN  GdPrv   

     MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  
1749         NaN       0      7    2009        WD         Normal  
1750        Shed      80      5    2010        WD         Normal  

[2 rows x 80 columns]

In [51]:
combined[1704:1753]

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape  \
1704  1705          50       RL         50.0     9350   Pave      Reg   
1706  1707          20       RL        110.0    14300   Pave      Reg   
1707  1708          50       RM         50.0     6000   Pave      Reg   
1708  1709         120       RL         43.0     3087   Pave      Reg   
1709  1710          20       RL         63.0     7560   Pave      Reg   
1710  1711          50       RM         63.0     4347   Pave      Reg   
1711  1712          20       RL         96.0    12444   Pave      Reg   
1712  1713          20       RL         85.0    11128   Pave      Reg   
1713  1714          80       RL         73.0     9735   Pave      Reg   
1714  1715          20       RL         83.0     9849   Pave      Reg   
1715  1716          20       RL         75.0    11625   Pave      Reg   
1716  1717          20       RL         70.0     8400   Pave      Reg   
1717  1718          50       RM         50.0     5925   Pave      Reg   
1718  1719          60       RL         85.0    11103   Pave      IR1   
1719  1720          20       RL         75.0    12518   Pave      IR1   
1720  1721          60       RL         88.0    12128   Pave      IR1   
1721  1722          85       RL         50.0     6723   Pave      Reg   
1722  1723          20       RL         75.0    13284   Pave      Reg   
1723  1724          60       RL         75.0     8970   Pave      IR1   
1724  1725          70       RL         79.0    11526   Pave      IR1   
1725  1726          85       RL         64.0     7301   Pave      Reg   
1726  1727          90       RM        110.0     8472   Grvl      IR2   
1727  1728         120       RL         53.0     4043   Pave      Reg   
1728  1729          20       RL         77.0    11500   Pave      IR1   
1729  1730          60       RL         59.0    11796   Pave      IR1   
1730  1731          50       RM         75.0     9000   Pave      Reg   
1731  1732          50       RM         52.0     6240   Pave      Reg   
1732  1733          50       RM         90.0    33120   Pave      IR3   
1733  1734          60       FV         72.0     8640   Pave      Reg   
1734  1735          60       RL         78.0     9978   Pave      IR1   
1735  1736          60       RL         82.0    16052   Pave      IR1   
1736  1737          20       RL         90.0    13377   Pave      Reg   
1737  1738          50       RM         60.0     5700   Pave      Reg   
1738  1739          20       RL         64.0     7314   Pave      Reg   
1739  1740         120       RL         53.0     4043   Pave      Reg   
1740  1741          50       RM         50.0     5925   Pave      Reg   
1741  1742          70       RM         60.0     6402   Pave      Reg   
1742  1743         120       RL         43.0     3136   Pave      IR1   
1743  1744          20       RL         71.0     9178   Pave      Reg   
1744  1745          30       RL         60.0     5400   Pave      Reg   
1745  1746          85       RL        150.0    14137   Pave      Reg   
1746  1747          60       RL         65.0    15426   Pave      IR1   
1747  1748          60       RL         75.0    11616   Pave      IR1   
1748  1749          20       RL         62.0    70761   Pave      IR1   
1749  1750          20       RL         75.0    17597   Pave      Reg   
1750  1751          80       RL         74.0     9620   Pave      Reg   
1751  1752          20       RL         78.0    10206   Pave      Reg   
1752  1753          50       RL         50.0     5000   Pave      Reg   
1753  1754          60       RL         75.0     9930   Pave      IR1   

     LandContour Utilities LotConfig  ... OpenPorchSF EnclosedPorch 3SsnPorch  \
1704         Bnk    AllPub    Inside  ...           0             0         0   
1706         HLS    AllPub    Inside  ...         150             0         0   
1707         Lvl    AllPub    Inside  ...          12             0         0   
1708         Lvl    AllPub    Ins

### 답도 행 자르기

In [52]:
print(targets.shape)

(1751,)


In [53]:
y = targets.drop([1705], axis=0)

In [54]:
print(y.shape)

(1750,)


In [55]:
y[1704:1708]

1704    112433.116092
1706    499600.821119
1707    125145.081485
1708    189993.872555
Name: SalePrice, dtype: float64

## one-hot-encoding

In [56]:
onehot_combined = pd.get_dummies(combined)

In [57]:
onehot_combined.head()

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0   1          20         75.0    21695            6            9       1988   
1   2          20         75.0    10530            6            5       1971   
2   3          20         88.0    12803            7            5       2002   
3   4          60        114.0     8314            7            5       1997   
4   5          20         78.0    10335            5            6       1968   

   YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  SaleType_ConLw  SaleType_New  \
0          2007       260.0       808.0  ...               0             0   
1          1971         0.0       282.0  ...               0             0   
2          2002        99.0       922.0  ...               0             0   
3          1998         0.0         0.0  ...               0             0   
4          1993         0.0       570.0  ...               0             0   

   SaleType_Oth  SaleType_WD  SaleCondition_Abnorml  SaleCondition_AdjLand  \
0             0            1                      0                      0   
1             0            1                      0                      0   
2             0            1                      0                      0   
3             0            1                      0                      0   
4             0            1                      0                      0   

   SaleCondition_Alloca  SaleCondition_Family  SaleCondition_Normal  \
0                     0                     0                     1   
1                     0                     0                     1   
2                     0                     0                     1   
3                     0                     0                     1   
4                     0                     0                     1   

   SaleCondition_Partial  
0                      0  
1                      0  
2                      0  
3                      0  
4                      0  

[5 rows x 275 columns]

## 원핫인코딩 후 Trina 과 Test 나누기

In [58]:
onehot_train = onehot_combined.iloc[:1750, :]
onehot_test = onehot_combined.iloc[1750:, :]

In [59]:
print(onehot_train.shape)
print(onehot_test.shape)

(1750, 275)
(1168, 275)


## 특성 중요도
- onehot_train : 결측치 처리(컬럼삭제, 행 1개삭제, 널값채우기) 후 원핫인코딩한 train
- onehot_test
- 랜덤포레스트회귀 모델
- 나무100개, random_state=7

In [78]:
from sklearn.ensemble import RandomForestRegressor

ranfo_reg = RandomForestRegressor(n_estimators=200, random_state=7)
ranfo_reg.fit(onehot_train, y)

fi = ranfo_reg.feature_importances_
importance_df = pd.DataFrame(fi, index=onehot_train.columns)
importance_df.sort_values(by=0, ascending=False)

0
OverallQual          6.288565e-01
GrLivArea            1.027249e-01
TotalBsmtSF          3.863824e-02
1stFlrSF             2.273811e-02
2ndFlrSF             2.048186e-02
BsmtFinSF1           1.968409e-02
FullBath             1.452663e-02
LotArea              1.436379e-02
BsmtQual_Ex          1.378110e-02
GarageCars           1.303523e-02
GarageArea           1.172396e-02
YearBuilt            8.833843e-03
YearRemodAdd         6.970259e-03
LotFrontage          5.996853e-03
OpenPorchSF          4.375727e-03
MasVnrArea           4.256840e-03
Id                   3.855417e-03
GarageYrBlt          3.789668e-03
BsmtUnfSF            3.776862e-03
WoodDeckSF           3.063547e-03
Fireplaces           2.909349e-03
MSSubClass           2.378228e-03
MSZoning_RM          2.338807e-03
OverallCond          2.284636e-03
TotRmsAbvGrd         2.256752e-03
CentralAir_N         1.780554e-03
KitchenQual_TA       1.703742e-03
MoSold               1.545359e-03
CentralAir_Y         1.401461e-03
GarageFinish_Unf     1.121530e-03
...                           ...
GarageQual_Po        1.353278e-06
Foundation_Stone     1.222769e-06
ExterCond_Po         1.063762e-06
SaleType_Con         9.599676e-07
Exterior2nd_AsphShn  9.512891e-07
HeatingQC_Po         9.382255e-07
Condition2_Feedr     8.619868e-07
Condition1_RRNe      7.533983e-07
GarageQual_Ex        6.813800e-07
SaleType_ConLw       6.312250e-07
RoofStyle_Shed       6.268603e-07
Heating_Wall         5.710982e-07
Electrical_FuseP     5.043929e-07
Heating_Floor        3.424971e-07
RoofMatl_ClyTile     3.171654e-07
Exterior2nd_CBlock   3.116367e-07
Condition2_RRNn      2.993193e-07
RoofMatl_Roll        2.850247e-07
Foundation_Wood      2.773692e-07
GarageCond_Ex        1.796540e-07
Exterior1st_AsphShn  1.224245e-07
Electrical_Mix       8.792566e-08
Condition2_Artery    6.875667e-08
Exterior1st_CBlock   5.724585e-08
RoofMatl_Membran     0.000000e+00
Exterior1st_ImStucc  0.000000e+00
Utilities_NoSeWa     0.000000e+00
Utilities_AllPub     0.000000e+00
Condition2_RRAe      0.000000e+00
Condition2_RRAn      0.000000e+00

[275 rows x 1 columns]

- 소수점예시 : 1.990832e-02 : 0.01990832

### 특성 선택

In [83]:
from sklearn.feature_selection import SelectFromModel                  # 특성선택해주는 라이브러리
selectModel = SelectFromModel(ranfo_reg, prefit=True, threshold=0.00007)
train_reduced = selectModel.transform(onehot_train)
test_reduced = selectModel.transform(onehot_test)
print(train_reduced.shape)
print(test_reduced.shape)

(1750, 142)
(1168, 142)


## Ridge 모델

In [84]:
from sklearn.model_selection import train_test_split                        
X_train, X_test, y_train, y_test = train_test_split(train_reduced, y, random_state=7) 

from sklearn.linear_model import Ridge
from sklearn.model_selection import cross_val_score

for a in range(1, 50):
    ridge = Ridge(alpha=a)
    result_ridge = cross_val_score(ridge, X_train, y_train, cv=5)
    print(a, result_ridge.mean())

1 0.8632880661552822
2 0.8650577096777831
3 0.8659393727902165
4 0.8664416466400221
5 0.8667390651588789
6 0.8669119054243228
7 0.8670033983237257
8 0.8670391930229165
9 0.8670355340493485
10 0.8670032068564144
11 0.8669496347165317
12 0.8668800747558072
13 0.866798337919566
14 0.8667072412013628
15 0.8666089016839725
16 0.8665049333095547
17 0.8663965818119248
18 0.8662848192173873
19 0.8661704112572977
20 0.8660539662505004
21 0.8659359710661387
22 0.8658168179303907
23 0.8656968246460295
24 0.8655762500081646
25 0.8654553056741354
26 0.8653341653858666
27 0.8652129721947512
28 0.8650918441660114
29 0.8649708789140611
30 0.8648501572330136
31 0.8647297460200921
32 0.864609700644013
33 0.8644900668730401
34 0.8643708824538419
35 0.8642521784082865
36 0.8641339801058088
37 0.8640163081523689
38 0.8638991791294567
39 0.8637826062132883
40 0.8636665996921984
41 0.8635511674029417
42 0.8634363150979064
43 0.8633220467564755
44 0.8632083648484512
45 0.8630952705581555
46 0.8629827639756578

alpha=32

## 학습하기

In [85]:
final_ridge = Ridge(alpha=11)
final_ridge.fit(X_train, y_train)
print(final_ridge.score(X_train, y_train))

0.9040901583904477


## 평가하기

In [86]:
final_ridge.fit(X_test, y_test)
print(final_ridge.score(X_test, y_test))

0.9621377127176487


## 예측하기

In [87]:
pre = final_ridge.predict(test_reduced)

## 저장

In [88]:
sub = pd.read_csv("sample_submission.csv", index_col=0)
sub["SalePrice"] = pre
sub.head()

SalePrice
Id                 
1752  220166.307789
1753  136591.494700
1754  213824.402872
1755  263451.530490
1756  175501.821821

In [89]:
sub.to_csv("house_Ridge1_(142).csv")